ROADMAP.md

```diff
@@ -0,0 +1,84 @@
+# Roadmap
+
+Contains subheaders about larger improvement efforts.
+
+## `rules_haskell` support
+
+Context: [`rules_haskell`](https://github.com/tweag/rules_haskell) supports
+building Haskell targets by extending Bazel, openly maintained by Tweag.
+It would be useful if `haskell-indexer` supported generating crossreferences for
+`rules_haskell` out of the box.
+
+### Goals
+
+Provide crossreference (as Kythe entries) for both `rules_haskell`-managed
+and `nixpkgs`-sourced compilations. Latter is used by `rules_haskell` for hackage
+imports, instead of hand-rolled source imports.
+
+Crossreference data should be serveable through Kythe's language server, ready to
+be consumed by emacs's lsp-mode (with modifications if needed).
+
+Bonus goal: based on local edits, either make partial index updates possible, or
+merge local updates with the static backing index at serving time.
+
+Possible future expansions:
+  * xref `proto-lens` generated Haskell protobuf code with proto sources.
+    (https://github.com/google/haskell-indexer/issues/15)
+  * xref Haskell FFI code with C code
+    (https://github.com/google/haskell-indexer/issues/18)
+
+### Tasks
+
+* Decide how compilations will be extracted/analysed.
+
+  * For quick productivity boost, run analysis directly, without extraction.
+    * `rules_haskell`: maybe just indicate indexing by passing some global flag
+      (if possible with Bazel).  Alternatively use `aspects`.
+    * `nixpkgs`: plumb a `should-index` flag to whatever helper is used to define
+      a Haskell package. Make `rules_haskell` pass this flag.
+    * Kythe entries will be written to a directory defined by the
+      `KYTHE_OUTPUT_DIRECTORY` env (convention).
+      
+  * Sync with Kythe/Bazel about merits of extraction, vs direct analysis.
```

Comment:
Kythe support for extracting Bazel rules already exist. Maybe we can get them open sources (TBH, might be the case that it's faster to write them from scratch then to open source them).
Maybe send an email to both Kythe and Bazel mailing lists? (if anything needs to be done internally at Google, feel free to put it on my plate)

Reply:
The internal extraction mostly dealt with finding and parsing back the `ghc` command line from the `SpawnAction` from the extracted `ExtraInfo`, also heuristics on deciding which inputs to capture. Since the external haskell rules are not the same as the internal, I don't expect too much code reuse.

What I want to discuss though, is doing away with `action_listener`, and rather using an aspect to do the extraction. That's something that could be interesting internally as well. That way you get better extraction, since guessing is not required, instead the compilation metadata can be used directly (internally Haddock and pals used that route already AFAIR).

Since the rules are different though, it's again not straightforward to make these converge. It would be a nice start to converge on the compilation metadata used (source files, extra flags, etc) per build rule. And eventually internal could just use the open-source rules_haskell ;)